libraries/freertos_plus/standard/crypto/src/iot_crypto.c

```diff
@@ -263,6 +278,101 @@ void CRYPTO_ConfigureThreading( void )
                                aws_mbedtls_mutex_unlock );
 }
 
+
+void CRYPTO_ConfigureDRBG( void )
+{
+    int lMbedResult = 0;
+
+    /* Initialize the entropy source and DRBG for the CRYPTO module */
+    mbedtls_entropy_init( &xDrbg.xMbedEntropyContext );
+    mbedtls_ctr_drbg_init( &xDrbg.xMbedDrbgCtx );
+
+    if( 0 != mbedtls_ctr_drbg_seed( &xDrbg.xMbedDrbgCtx,
+                                    mbedtls_entropy_func,
+                                    &xDrbg.xMbedEntropyContext,
+                                    NULL,
+                                    0 ) )
+    {
+        CRYPTO_PRINT( ( "Failed to seed DRBG" ) );
+    }
+    else
+    {
+        xDrbg.xIsInitialized = CK_TRUE;
+    }
+}
+
+
+
+int mbedtls_hardware_poll( void * data,
+                           unsigned char * output,
+                           size_t len,
+                           size_t * olen )
+{
+    int lStatus = MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+    CK_RV xSession = CK_INVALID_HANDLE;
+    CK_RV xResult = CKR_OK;
+    CK_FUNCTION_LIST_PTR pxFunctionList = NULL;
+    CK_BBOOL xSessionOpened = CK_FALSE;
+
+    xResult = C_GetFunctionList( &pxFunctionList );
+
+    if( xResult == CKR_OK )
+    {
+        xResult = xInitializePkcs11Session( &xSession );
+    }
+
+    if( xResult == CKR_OK )
+    {
+        xSessionOpened = CK_TRUE;
+        xResult = pxFunctionList->C_GenerateRandom( xSession, output, len );
+    }
+
+    if( xSessionOpened == CK_TRUE )
+    {
+        xResult = pxFunctionList->C_CloseSession( xSession );
+    }
+
+    if( xResult == CKR_OK )
+    {
+        lStatus = 0;
+        *olen = len;
+    }
+
+    return lStatus;
+}
+
+
+
+int CRYPTO_GetRandomBytes( void * pvCtx,
+                           uint8_t * pRandomBytes,
+                           size_t length )
+{
+    ( void ) pvCtx; /*lint !e9087 !e9079 Allow casting void* to other types. */
+
+    int lResult = 0;
+
+    if( ( NULL == pRandomBytes ) ||
+        ( length == 0 ) )
+    {
+        CRYPTO_PRINT( ( "ERROR: Invalid inputs to CRYPTO_GetRandomBytes \r\n" ) );
+        lResult = CRYPTO_ERROR_RNG;
+    }
+
+    if( lResult == 0 )
+    {
+        lResult = mbedtls_ctr_drbg_random( &xDrbg.xMbedDrbgCtx, pRandomBytes, length );
```

Comment:
Does there need to be a test of xIsInitialized here?

Reply:
yes thank you fixed.